pkg/syncer/shared/namespace.go

```diff
@@ -31,14 +35,25 @@ const (
 // NamespaceLocator stores a logical cluster and namespace and is used
 // as the source for the mapped namespace name in a physical cluster.
 type NamespaceLocator struct {
-	LogicalCluster logicalcluster.Name `json:"logical-cluster"`
-	Namespace      string              `json:"namespace"`
+	Workspace          logicalcluster.Name `json:"workspace,omitempty"`
+	WorkloadCluster    logicalcluster.Name `json:"workloadCluster,omitempty"`
+	WorkloadClusterUID types.UID           `json:"workloadClusterUID,omitempty"`
+	Namespace          string              `json:"namespace"`
+}
+
+func NewNamespaceLocator(workspace, workloadClusterWorkspace logicalcluster.Name, workloadClusterUID types.UID, workloadLogicalClusterName, upstreamNamespace string) NamespaceLocator {
+	return NamespaceLocator{
+		Workspace:          workspace,
+		WorkloadCluster:    workloadClusterWorkspace.Join(workloadLogicalClusterName),
+		WorkloadClusterUID: workloadClusterUID,
+		Namespace:          upstreamNamespace,
+	}
 }
 
 func LocatorFromAnnotations(annotations map[string]string) (*NamespaceLocator, error) {
-	annotation := annotations[NamespaceLocatorAnnotation]
-	if len(annotation) == 0 {
-		return nil, nil
+	annotation, ok := annotations[NamespaceLocatorAnnotation]
+	if !ok || len(annotation) == 0 {
+		return nil, fmt.Errorf("no namespace locator annotation found")
```

Comment:
not a good signature. This is not an error, i.e. some exceptional state, but something normal. I.e. either the old semantics are good, or `return nil, false, nil` vs. `return locator, true, nil`.